crates/astria-gossipnet/src/network.rs

```diff
@@ -131,43 +136,119 @@ impl Network {
 
         // build a gossipsub network behaviour
         let gossipsub = gossipsub::Behaviour::new(
-            gossipsub::MessageAuthenticity::Signed(local_key),
+            gossipsub::MessageAuthenticity::Signed(keypair),
             gossipsub_config,
         )
         .map_err(|e| eyre!("failed to create gossipsub behaviour: {}", e))?;
 
         let mut swarm = {
-            #[cfg(feature = "mdns")]
-            let mdns = mdns::tokio::Behaviour::new(mdns::Config::default(), local_peer_id)?;
+            let kademlia = if self.with_kademlia {
+                let mut cfg = KademliaConfig::default();
+                cfg.set_query_timeout(Duration::from_secs(5 * 60));
+                let store = MemoryStore::new(local_peer_id);
+                Some(Kademlia::with_config(local_peer_id, store, cfg))
+            } else {
+                None
+            };
+
+            let mdns = if self.with_mdns {
+                Some(mdns::tokio::Behaviour::new(
+                    mdns::Config::default(),
+                    local_peer_id,
+                )?)
+            } else {
+                None
+            };
+
             let behaviour = GossipnetBehaviour {
+                identify: identify::Behaviour::new(identify::Config::new(
+                    GOSSIPNET_PROTOCOL_ID.into(),
+                    public_key,
+                )),
                 gossipsub,
-                #[cfg(feature = "mdns")]
-                mdns,
                 ping: ping::Behaviour::default(),
+                mdns: mdns.into(),
+                kademlia: kademlia.into(),
             };
             SwarmBuilder::with_tokio_executor(transport, behaviour, local_peer_id).build()
         };
 
-        let listen_addr = format!("/ip4/0.0.0.0/tcp/{}", port);
+        let listen_addr = format!("/ip4/0.0.0.0/tcp/{}", self.port);
         swarm.listen_on(listen_addr.parse()?)?;
```

Comment:
i know not touched these lines in this PR, but if we are at it...: error context on both `?`